File: Scala/algorithmic_toolbox/src/main/scala-2.12/introduction/PartialSumOfFibonacciNumbers.scala
package introduction

import scala.annotation.tailrec
import utils.RichNumber._

object PartialSumOfFibonacciNumbers {
  private val Modulus: Int = 10

  private def convertToLongList(line: String): List[Long] = line.split(" ").map(_.toLong).toList

  private def calcPisanoPeriod(modulus: Int): Long = {
    @tailrec
    def loop(a: Int, b: Int, p: Long): Long = {
      if (a == 0 && b == 1) p
      else loop(b, (a + b) % modulus, p + 1)
    }
    loop(a = 1, b = 1, p = 1L)
  }

  private def calcFibonacciModulo(n: Long, modulus: Int): Int =
    (0L until n).foldLeft((0, 1)){ case ((a, b), _) => (b, (a + b) % modulus) }._1

  def calcLastDigitOfPartialSum(m: Long, n: Long): Int = {
    val p: Long = calcPisanoPeriod(Modulus)
    val lastDigitOfPrefixSum: Int = (calcFibonacciModulo((m + 1) % p, Modulus) - 1) % Modulus
    val lastDigitOfFullSum: Int = (calcFibonacciModulo((n + 2) % p, Modulus) - 1) % Modulus
    (lastDigitOfFullSum - lastDigitOfPrefixSum).mod(Modulus)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val List(m, n): List[Long] = convertToLongList(reader.next())
    val result: Int = calcLastDigitOfPartialSum(m, n)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/introduction/partial_sum_of_fibonacci_numbers.ml
let modulus = 10


let modulo a b =
    let result = a mod b in
    if result >= 0 then result
    else result + b


let calc_pisano_period modulus =
    let rec inner a b p =
        if a = 0 && b = 1 then p
        else inner b ((a + b) mod modulus) (p + 1)
    in inner 1 1 1


let calc_fibonacci_modulo n modulus =
    let rec inner k a b =
        if k < 1 then a
        else inner (k - 1) b ((a + b) mod modulus)
    in inner n 0 1


let calc_last_digit_of_partial_sum m n =
    let p = calc_pisano_period modulus in
    let last_digit_m_minus_one = modulo ((calc_fibonacci_modulo ((m + 1) mod p) modulus) - 1) modulus in
    let last_digit_n = modulo ((calc_fibonacci_modulo ((n + 2) mod p) modulus) - 1) modulus
    in modulo (last_digit_n - last_digit_m_minus_one) modulus


let () =
    let (m, n) = Scanf.scanf "%d %d" (fun x y -> (x, y)) in
    print_int (calc_last_digit_of_partial_sum m n);
    print_newline()
